apm-sniffer/apm-toolkit-activation/apm-toolkit-logback-1.x-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/log/logback/v1/x/log/GRPCLogAppenderInterceptor.java

```diff
@@ -0,0 +1,93 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.toolkit.activation.log.logback.v1.x.log;
+
+import java.lang.reflect.Method;
+import java.util.Objects;
+
+import org.apache.skywalking.apm.agent.core.boot.ServiceManager;
+import org.apache.skywalking.apm.agent.core.conf.Config;
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+import org.apache.skywalking.apm.agent.core.remote.LogReportServiceClient;
+import org.apache.skywalking.apm.network.common.v3.KeyStringValuePair;
+import org.apache.skywalking.apm.network.logging.v3.LogData;
+import org.apache.skywalking.apm.network.logging.v3.LogDataBody;
+import org.apache.skywalking.apm.network.logging.v3.LogTags;
+import org.apache.skywalking.apm.network.logging.v3.TextLog;
+import org.apache.skywalking.apm.network.logging.v3.TraceContext;
+
+import ch.qos.logback.classic.spi.ILoggingEvent;
+
+public class GRPCLogAppenderInterceptor implements InstanceMethodsAroundInterceptor {
+
+    private LogReportServiceClient client;
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                             MethodInterceptResult result) throws Throwable {
+        if (Objects.isNull(client)) {
+            client = ServiceManager.INSTANCE.findService(LogReportServiceClient.class);
+            if (Objects.isNull(client)) {
+                return;
+            }
+        }
+        ILoggingEvent event = (ILoggingEvent) allArguments[0];
+        if (Objects.nonNull(event)) {
+            client.produce(transform(event));
+        }
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                              Object ret) throws Throwable {
+        return ret;
+    }
+
+    @Override
+    public void handleMethodException(EnhancedInstance objInst, Method method, Object[] allArguments,
+                                      Class<?>[] argumentsTypes, Throwable t) {
+
+    }
+
+    private LogData transform(ILoggingEvent event) {
+        return LogData.newBuilder()
+                .setTimestamp(event.getTimeStamp())
+                .setService(Config.Agent.SERVICE_NAME)
+                .setServiceInstance(Config.Agent.INSTANCE_NAME)
+                .setTraceContext(TraceContext.newBuilder()
+                        .setTraceId(ContextManager.getGlobalTraceId())
+                        .setSpanId(ContextManager.getSpanId())
+                        .setTraceSegmentId(ContextManager.getSegmentId())
```

Comment:
@EvanLjp As we are always sending this, should we still keep trace id as empty string, if there is no tracing context? What do you think?

Reply:
when no tracing context, the traceId maybe `N/A` rather than an empty string.

Reply:
Do you want to send `N/A` always?

Reply:
we can reduce the cost on the network, @arugal I think the oap server query mode is compatible with this way. Is that right?

Reply:
https://github.com/apache/skywalking/blob/9021e6791229e8dd81ffaf7b77cb9f567723d0b6/oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/listener/RecordAnalysisListener.java#L85-L92

Now `N/A` is not filtered, I will submit a PR fix.

Reply:
Don't fix for now. I am still feeling we should not send them.

Reply:
I am thinking, either core return empty even in tracing context, or the plugin filters this. After all, my key concern is network cost.

Reply:
@EvanLjp What do you prefer? Filter `N/A` here?

Reply:
@wu-sheng  I think the trace context should be removed, but the log content still should be sent.

Reply:
It is fine to send the logs, we at least have the service and service instance name.

Reply:
I mean the segment id and span id only.

Reply:
@EvanLjp @elk-g Let's finalize this, others are good now. There are `N/A`, `Ignored_Trace`, ``(EMPTY_STRING`) as trace id and segment id.  -1 as span id, which all means no available trace context. Could we filter these out? At least base on `trace_id`?